pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+                  indicates how the block was constructed.
+                  It is used to differentiate between the possible future
+                  changes to the DataMap format.
+                - block size: 32-bit integer that represents the total size of
+                  block including the header and the data.
+                - Number of Scalars: number of scalar variables
+                - Scalar data: the scalar data of the record.
+                  Please see DmapScalar for more information on scalars.
+                - Number of arrays: number of array variables
+                  Please see DmapArray for more information on arrays.
+            """
+            # This is an unused variable but is need to move the cursor to the
+            # next offset.
+            # TODO: Possible check that uses the encoding identifier
+            # encoding_identifier = self.read_data('i',4)
+            self.cursor += 4
+            block_size = self.read_data('i',4)
+            self.zero_check(block_size, "block size")
+
+            total_block_size += block_size
+            self.bytes_check(total_block_size, "total block size",
+                             self.dmap_end_bytes, "total bytes in file")
+
+            # why minus 2?
+            self.cursor = self.cursor + block_size - 2 * DMAP_DATA_TYPES[INT][1]
+
+        if total_block_size != self.dmap_end_bytes:
+            message = "Error: Initial integrity check shows"\
+                    " total block size: {total_size} < end bytes {end_bytes}."\
+                    " Cursor: {cursor}."\
+                    " Data is likely corrupted"\
+                    "".format(total_size=total_block_size,
+                              end_bytes=self.dmap_end_bytes,
+                              cursor=self.cursor)
+            raise pydmap_exceptions.DmapDataError(self.date_file, message)
+        self.cursor = 0
+
+    def read_records(self) -> collections.deque:
+        """
+        This method reads the records from the dmap file/stream passed
+        into the instance.
+
+
+        Return
+        -------
+        __dmap_records : collections.Deque
+                Deque list of DmapRecords (ordered dictionary)
+
+
+        See Also
+        --------
+        DmapRecord : dmap record's data structure
+        DmapScalar : dmap record's scalar data structure
+        DmapArray  : dmap record's array data structure
+
+        See DEVELOPER_README.md for more information DmapRecords data structure.
+        """
+
+        # read bytes until end of byte array
+        pydarn_logger.debug("Reading DMAP records")
+        while self.cursor < self.dmap_end_bytes:
+            new_record = self.read_record()
+            self.__dmap_records.append(new_record)
+
+        self.bytes_check(self.cursor, "cursor",
+                         self.dmap_end_bytes, "total bytes in the file")
+        return self.__dmap_records
+
+    def read_record(self) -> collections.OrderedDict:
```

Comment:
Actually, it depends, IO alone for parallelizing is a difficult challenge because typically you want to read things in order and write things in order and share a common file. Then you have to ask when parallelization is worth it? currently, on my computer alone it takes 1.945 s to read a 42MB RAWACF. If I add multiprocessing (mp) into the mix, you might not get any decrease in time because of the overhead to call mp and then create the processes (might actually increase the time). Also, you might need to know how many processes are available for the user's machine. Typically best practice is to optimize single process method first then look at concurrency if possible, and you have something to gain from it. 

However, I can see the benefit in parallelizing to read large full day FITACF files. Parallelization is possible to do since there are defined block sizes and with how modular the code is right now, it would be a couple of lines of code to change. One caveat is we probably would want the list to reflect the DMAP file so we would need to also spend some time on re-sorting the list. 

Good idea but more work needs to be looked into with how big does the file need to be to get a performance increase for DmapRead alone finally. Especially if @asreimer fixes cdmap then there is some exciting performance testing to be done :) 

In the testing branch, I am actually going to make a performance suite for testing performance of code :) (Request made by @asreimer)

Again out of scope and future want. So I will come back to this in the future. The main goal is to get the same capabilities as DaVitpy and then start adding in better performance methods and other features.

Reply:
Yeah I'm not 100% sure this is feasible. I do know that Pool() automatically determines number of available processors and that mp has a thing that will automatically preserve input ordering so that you don't need to rearrange, but I don't know about performance overheads.